### PR TITLE
Fix multiprocessing logging

### PR DESCRIPTION
**Pull request by daniel-wer**
See https://github.com/scalableminds/voxelytics/pull/2641

### CODE REVIEW

File: cluster_tools/remote.py [hunk 1]
```diff
@@ -92,15 +92,20 @@ def worker(workerid, job_array_index, cfut_dir):
...
         )
-        meta_data["logging_setup_fn"](log_file_path)
-        logging.debug("Using supplied logging_setup_fn to setup logging.")
+        meta_data["logging_remote_setup_fn"](log_file_path)
+        logging.debug("Using supplied logging_remote_setup_fn to setup logging.")
+    elif "logging_multiprocessing_setup_fn" in meta_data:
+        meta_data["logging_multiprocessing_setup_fn"]()
+        logging.debug(
+            "Using supplied logging_multiprocessing_setup_fn to setup logging."
+        )
```
Comment (philippotto) [inline cluster_tools/remote.py:108]: I only moved the logging code one line up each. That way, stdout will already state what's going on, even if the user-supplied setup functions fail. I think, this is a tad better, but could also be a matter of taste.

```suggestion
        logging.debug("Using supplied logging_remote_setup_fn to setup logging.")
        meta_data["logging_remote_setup_fn"](log_file_path)
    elif "logging_multiprocessing_setup_fn" in meta_data:
        logging.debug(
            "Using supplied logging_multiprocessing_setup_fn to setup logging."
        )
        meta_data["logging_multiprocessing_setup_fn"]()
```

**Review by philippotto — APPROVED** [review]
Great stuff! So, before that PR the logging setup was only called for slurm/pbs and now there are two different ways of setting up logging (locally vs. remote) where the local way is only used if the process wasn't forked. Did I get this right? Maybe it would be a bit more generic, if the local-setup-function was **always** called and the caller is responsible for deciding whether something needs to be done (i.e., check for the start-method). But it probably doesn't really matter..

File: cluster_tools/__init__.py [hunk 2]
```diff
@@ -72,15 +81,22 @@ def submit(self, *args, **kwargs):
         else:
             submit_fn = super().submit
 
+        call_stack = []
```
[the remark this answers is not in the document]
Comment (daniel-wer) [reply cluster_tools/__init__.py:80]: Does the comment I added explain it well? It got a little verbose :see_no_evil:

File: cluster_tools/schedulers/cluster_executor.py [hunk 3]
```diff
@@ -84,12 +75,14 @@ def __init__(
 
         self.meta_data = {}
         assert not (
-            "logging_config" in kwargs and "logging_setup_fn" in kwargs
-        ), "Specify either logging_config OR logging_setup_fn but not both at once"
+            "logging_config" in kwargs and "logging_remote_setup_fn" in kwargs
```
[the remark this answers is not in the document]
Comment (daniel-wer) [reply cluster_tools/schedulers/cluster_executor.py]: I'll resolve this comment as it is no longer applicable. I reverted the naming change.

**Comment by daniel-wer** [comment]
@philippotto Following our discussion, I moved the multiprocessing logging handler code to the cluster tools which now take care that the logging just works:tm: . The PR in voxelytics can be closed and I was able to remove a lot of the interface complexity. I like this version a lot :)

**Comment by daniel-wer** [comment]
I reran the predict task of a regression test with the new cluster_tools version.

Output before: `less /conndata/georgwie/artifacts/segem_intra_extra_regressiontest_master_2021-07-23/predict/prediction_v1__b3085aedb9-v1/cfut/slurmpy.22336106_3.log.stdout`

```

2021-07-26 14:38:02.685459: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
trying to read:  artifacts/segem_intra_extra_regressiontest_master_2021-07-23/predict/prediction_v1__b3085aedb9/cfut/cfut.in.J0q3wsVB7cJXlqc5MPcPVkXWjXNGG5ET_3.pickle
working dir:  /tmpscratch/jonathan/regression_tests/voxelytics
2021-07-26 14:38:20 PID=5186 DEBUG    root Using supplied logging_setup_fn to setup logging.
2021-07-26 14:38:20 PID=5186 INFO     root Job computation started (jobid=22336110, workerid_with_idx=J0q3wsVB7cJXlqc5MPcPVkXWjXNGG5ET_3).
2021-07-26 14:38:20 PID=5186 DEBUG    voxelytics.commons.resumable_executor Running step for chunk BoundingBo...
2021-07-26 14:38:20 PID=5186 DEBUG    voxelytics.commons.resumable_executor Looking for checkpoint: artifacts/segem_intra_extra_regressiontest_master_2021-07-23/predict/prediction_v1__b3085aedb9/resumable_executions/main/step/20,256,256,236,124,86/checkpoint. Exists: False
/tmpscratch/jonathan/regression_tests/voxelytics/environments/miniconda/envs/vx_regressiontest-py38/lib/python3.8/site-packages/networkit/plot.py:13: ReducedFunctionalityWarning: WARNING: module 'seaborn' is not installed, plotting functionality will be limited
  warnings.warn("WARNING: module 'seaborn' is not installed, plotting functionality will be limited",
2021-07-26 14:38:20 PID=5186 INFO     voxelytics.connect.abstract_tasks.distributable_wkw_task Running step for sub-chunk BoundingBox(topleft=(20, 256, 256), size=(236, 124, 86))
2021-07-26 14:38:20 PID=5186 INFO     voxelytics.connect.tools.predict Reading in datasource
2021-07-26 14:38:20 PID=5186 INFO     voxelytics.connect.tools.predict Using prediction output_shape: (259, 259, 257)
2021-07-26 14:38:20 PID=5186 INFO     voxelytics.connect.tools.predict initializing inferrer
2021-07-26 14:38:20 PID=5186 DEBUG    WkwDataReader Loading data into memory: segem_cortex_test
2021-07-26 14:38:20 PID=5186 DEBUG    WkwDataReader Data volume shape: (300, 300, 274)
2021-07-26 14:38:20.759891: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-07-26 14:38:20.761518: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1
2021-07-26 14:38:21.228161: E tensorflow/stream_executor/cuda/cuda_driver.cc:328] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2021-07-26 14:38:21.228233: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (gaba014): /proc/driver/nvidia/version does not exist
2021-07-26 14:38:21.229056: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-07-26 14:38:21.229657: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-07-26 14:38:21.308325: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:196] None of the MLIR optimization passes are enabled (registered 0 passes)
```

Output after: `less /conndata/georgwie/artifacts/segem_intra_extra_regressiontest_master_2021-07-23/predict/prediction_v1__b3085aedb9-v2/cfut/slurmpy.22588630_2.log.stdout`

```

2021-08-04 13:38:08.781534: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
trying to read:  artifacts/segem_intra_extra_regressiontest_master_2021-07-23/predict/prediction_v1__b3085aedb9/cfut/cfut.in.QW49smafwZysaWEouysoQiii0DHKcRtt_2.pickle
working dir:  /gaba/u/danielwe/code/voxelytics
2021-08-04 13:38:12 PID=25846 INFO     root Job computation started (jobid=22588649, workerid_with_idx=QW49smafwZysaWEouysoQii2021-08-04 13:38:08.781534: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic librar
y libcudart.so.11.0
trying to read:  artifacts/segem_intra_extra_regressiontest_master_2021-07-23/predict/prediction_v1__b3085aedb9/cfut/cfut.in.Q
W49smafwZysaWEouysoQiii0DHKcRtt_2.pickle
working dir:  /gaba/u/danielwe/code/voxelytics
2021-08-04 13:38:12 PID=25846 INFO     root Job computation started (jobid=22588649, workerid_with_idx=QW49smafwZysaWEouysoQii
i0DHKcRtt_2).
/u/danielwe/code/voxelytics/environments/miniconda/envs/vx-py38/lib/python3.8/site-packages/networkit/plot.py:13: ReducedFunctionalityWarning: WARNING: module 'seaborn' is not installed, plotting functionality will be limited
  warnings.warn("WARNING: module 'seaborn' is not installed, plotting functionality will be limited",
2021-08-04 13:38:12 PID=25846 INFO     voxelytics.connect.abstract_tasks.distributable_wkw_task Running step for sub-chunk BoundingBox(topleft=(20, 256, 8), size=(236, 124, 248))
2021-08-04 13:38:12 PID=25846 INFO     voxelytics.connect.tools.predict Reading in datasource
2021-08-04 13:38:12 PID=25846 INFO     voxelytics.connect.tools.predict Using prediction output_shape: (259, 259, 257)
2021-08-04 13:38:12 PID=25846 INFO     voxelytics.connect.tools.predict initializing inferrer
2021-08-04 13:38:13.123219: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-08-04 13:38:13.124833: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1
2021-08-04 13:38:13.133158: E tensorflow/stream_executor/cuda/cuda_driver.cc:328] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2021-08-04 13:38:13.133204: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:169] retrieving CUDA diagnostic information for host: gaba007
2021-08-04 13:38:13.133219: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:176] hostname: gaba007
2021-08-04 13:38:13.133330: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:200] libcuda reported version is: 450.102.4
2021-08-04 13:38:13.133386: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:204] kernel reported version is: 450.102.4
2021-08-04 13:38:13.133401: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:310] kernel version seems to match DSO: 450.102.4
2021-08-04 13:38:13.134201: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-08-04 13:38:13.207646: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:196] None of the MLIR optimization passes are enabled (registered 0 passes)
2021-08-04 13:38:13.283362: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 2793215000 Hz
2021-08-04 13:38:29 PID=25846 INFO     voxelytics.connect.data.data_readers Inference of (3,) took 17.200199365615845s
2021-08-04 13:38:30 PID=25846 INFO     voxelytics.connect.tools.predict Percentiles 2 and 98: (-1.9077983856201173, 8.695617485046387)
2021-08-04 13:38:30 PID=25846 INFO     voxelytics.connect.abstract_tasks.distributable_wkw_task Finished task chunk 1 / 1
2021-08-04 13:38:30 PID=25846 INFO     voxelytics.connect.abstract_tasks.distributable_wkw_task Save wkw artifact to artifacts/segem_intra_extra_regressiontest_master_2021-07-23/predict/prediction_v1__b3085aedb9/prediction/prediction
2021-08-04 13:38:30 PID=25846 INFO     voxelytics.commons.wkw_utils write at artifacts/segem_intra_extra_regressiontest_master_2021-07-23/predict/prediction_v1__b3085aedb9/prediction/prediction/1
2021-08-04 13:38:30 PID=25846 INFO     voxelytics.connect.abstract_tasks.distributable_wkw_task Writing took 0.24730229377746582s
2021-08-04 13:38:31 PID=25846 INFO     root Job computation completed.
```

As you can see in the old version, the last messages (from the actual executed task code) are missing which is no longer the case with the logging fix.

File: cluster_tools/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (philippotto) [reply cluster_tools/__init__.py:80]: Perfect! This should make the code easy to understand in the future :)

**Review by philippotto — APPROVED** [review]
Excellent! Looks way better than the previous iteration :+1: The code makes sense and since you tested it, it should be good to go :100:

**Pull request merged**